[로또] 1단계 미션 제출합니다.

nimunsang opened the pull request.

Hue9010 commented on src/main/java/domain/LottoTickets.java:
```diff
@@ -0,0 +1,64 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class LottoTickets {
```
> List<LottoTicket>에 대한 일급 컬렉션으로 LottoTickets를 만들어서 사용하는데, 이 LottoTickets에 대한 정보를 전달할 때에도 따로 객체를 만들어서 전달해줘야 할까요? (예를들어 LottoTicketsDto) 현재 LottoTickets은 데이터를 가공해서 정보를 추출해주는 책임만 가지고 있습니다.

`LottoTickets에 대한 정보를 전달할 때`가 View를 말씀 하시는 거라면 만들어줘야 합니다. 정확히는 LottoTickets의 List<LottoTicket>도 Dto 만들어야 합니다.

```java
public class LottoTicketsDto {
    private List<LottoTicketsDto> lottoTickets;
```

이번 미션처럼 콘솔이 아니라 웹이였다면 아래와 같은 네이밍의 DTO를 만들었을 것 같네요.

```java
public class LottoTicketsResponse {
    private List<LottoTicketsResponse> lottoTickets;
```

DTO는 Data Transfer Object 뜻 그대로 데이터 전달만을 목적으로 하는 객체인데 터너(이렇게 부르는게 맞나요? 😋 )도 말씀하신 것 처럼 LottoTickets는 Lotto List의 관리를 하고 있습니다. 이미 DTO의 정의를 벗어 났습니다. 그래서 질문 의도에 맞춰 답변 드리면 DTO를 만드는게 맞습니다.
그러면 이번에 DTO를 만들어야 하냐면 No입니다. 우선 DTO를 만들어 얻을 수 있는 이점이 없고 그에 비해 지나치게 프로젝트가 복잡해지기 때문입니다.

워낙 옛날 글이라 요즘은 다른 견해를 가지고 계실 수도 있는데 [혹시 DTO(VO) 작성하시나요?](https://www.slipp.net/questions/22)라는 글에서 익숙하신 분이 보이네요. 🙄

sumniy replied:
Dto에 대한 개념이 잘 잡힌 것 같습니다 감사합니다! (터너 맞습니다 👌)

Hue9010 commented on src/main/java/domain/LottoController.java:
```diff
@@ -0,0 +1,30 @@
+        LottoOutputView.printYield(amount, winningInfo.getSumPrize());
+    }
+
+    private static LottoTickets createLottoTickets(int ticketCount) {
+        LottoTickets lottoTickets = LottoService.generateRandomLottoTickets(ticketCount);
+        LottoOutputView.printTickets(lottoTickets.getLottoTicketsInfo());
+        return lottoTickets;
+    }
```
> InputView를 사용했고, View 클래스는 입력에 대한 책임만 있어야 한다고 생각해서 입력받은 데이터만 넘겨주고, 실제 데이터를 이용한 객체 생성은 컨트롤러가 담당하게 했습니다. 그러다보니 컨트롤러의 코드가 지저분해졌는데, 가독성을 높이게 하기 위해서 InputView 클래스에 객체를 생성해서 반환하는 책임을 가지게 하는 방법은 클린 코드에 적합하지 않은 방법일까요?

[🚀 2단계 - 자동차 경주 리팩토링](https://edu.nextstep.camp/s/bb4PXLji/ls/W1LFM86E)에 있는 그림을 다시 한번 보죠!

![mvc](https://firebasestorage.googleapis.com/v0/b/nextstep-real.appspot.com/o/lesson-attachments%2F-L9D5L4Xj_6xjTfOTRks%2Fdomain.PNG?alt=media&token=5083069f-e922-4f0e-95a4-d3fb9fb95224)

클린코드는 넓은 의미를 가지니 범위를 좁혀 얘기하면 MVC 패턴에 맞지 않습니다. MVC 패턴을 적용했지만 MVC 패턴을 제대로 적용하지 못한 점에서 클린 하지 못하다고 볼 수 있겠네요.

위 그림 위에
> MVC 패턴 기반으로 리팩토링해 view 패키지의 객체가 domain 패키지 객체에 의존할 수 있지만, domain 패키지의 객체는 view 패키지 객체에 의존하지 않도록 구현한다

라는 언급이 있는데 이는 패키지 의존성에 한정 된 얘기입니다.(View에서 Model을 화면에 출력해야하니까 허용)

 View가 지금의 자바 환경이 아닌 웹 페이지로 생각해보시면 조금 더 이해가 빠를 수도 있을 것 같네요. 😉 View에서 객체 생성을 하게 되면 View를 콘솔이 아닌 웹페이지로 변환이 가능할까요?

Hue9010 replied:
Controller가 domain 패키지에 위치하고 있네요. 알맞은 패키지를 만들어서 옮겨주세요. 😉

sumniy replied:
웹 페이지로 생각해보니 바로 이해가 갔습니다 ㅎㅎ 감사합니다 😀

Hue9010 commented on src/test/java/domain/LottoTicketTest.java:
```diff
@@ -0,0 +1,51 @@
+
+public class LottoTicketTest {
+
+    @Test
+    public void create() {
+        LottoTicket lottoTicket = LottoTicket.ofIntegerNumber(Arrays.asList(1, 2, 3, 4, 5, 45));
+        assertThat(lottoTicket).isEqualTo(LottoTicket.ofIntegerNumber(Arrays.asList(1, 2, 3, 4, 5, 45)));
+    }
+
+    @Test
+    public void isInValidLength() {
+        assertThatExceptionOfType(InvalidLottoNumberException.class)
+                .isThrownBy(() -> LottoTicket.ofIntegerNumber(Arrays.asList(1, 2, 3, 4, 5)));
+        assertThatExceptionOfType(InvalidLottoNumberException.class)
+                .isThrownBy(() -> LottoTicket.ofIntegerNumber(Arrays.asList(1, 2, 3, 4, 5, 6, 7)));
+        assertThatExceptionOfType(InvalidLottoNumberException.class)
+                .isThrownBy(() -> LottoTicket.ofIntegerNumber(Arrays.asList()));
+        assertThatExceptionOfType(InvalidLottoNumberException.class)
+                .isThrownBy(() -> LottoTicket.ofIntegerNumber(null));
+    }
```
예외가 발생하고 있는 케이스까지 꼼꼼하게 잘 해주셨네요. 👍

Hue9010 reviewed (changes requested):
안녕하세요. 이번 리뷰를 맡은 휴입니다. 😄 
객체 분리도 잘하셨고 전반적으로 군더더기 없이 깔끔히 잘 짜주셨네요. 👍 
질문 주신 사항 답변과 피드백 남겼는데 확인하시고 궁금하신 사항 있으면 문의해주세요.

sumniy commented on src/main/java/domain/LottoTicket.java:
```diff
@@ -0,0 +1,75 @@
+
+    private final List<LottoNumber> numbers;
+
+    public static LottoTicket ofLottoNumber(List<LottoNumber> numbers) {
+        return new LottoTicket(numbers);
+    }
```
말씀 듣고 다시 보니 확실히 넘겨주는 타입으로 알 수 있다는 생각이 드네요. 알려주셔서 감사합니다 🤗

sumniy commented on src/main/java/domain/LottoTicket.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    public List<Integer> getLottoTicketInfo() {
+        return numbers.stream().map(LottoNumber::getNumber).collect(Collectors.toList());
```
앗 ㅎㅎ 감사합니다

sumniy commented on src/main/java/domain/WinningInfo.java:
```diff
@@ -0,0 +1,42 @@
+
+public class WinningInfo {
+
+    private final List<Long> PRIZE = Arrays.asList(5000l, 50000l, 1500000l, 30000000l, 2000000000l);
+
+    private final List<Integer> info;
```
거의 하드코딩에 가까운 방법이어서 고민하고 있던 부분인데, 좋은 방법 알려주셔서 감사합니다! 훨씬 관리하기 편해지고 가독성이 좋아질 것 같네요!

sumniy commented on src/main/java/view/LottoOutputView.java:
```diff
@@ -0,0 +1,58 @@
+
+    public static void printYield(Amount amount, Long sum) {
+        double yield = (double) sum / amount.getAmount();
+        System.out.println(String.format("총 수익률은 %.2f입니다.", yield));
```
헉 그 부분은 모르고 있었네요 🙇‍♂️

sumniy commented on src/main/java/domain/LottoController.java:
```diff
@@ -0,0 +1,30 @@
+        LottoWinningNumber lottoWinningNumber = new LottoWinningNumber(lottoWinningTicket, bonusNumber);
+        WinningInfo winningInfo = lottoTickets.getWinningInfo(lottoWinningNumber);
+        LottoOutputView.printResult(winningInfo);
+        LottoOutputView.printYield(amount, winningInfo.getSumPrize());
```
그렇게 하는편이 뷰에서 연산하는 로직도 뺄 수 있겠네요

sumniy commented:
피드백 반영했습니다. 리뷰 부탁드립니다!

Hue9010 commented on src/main/java/domain/LottoTickets.java:
```diff
@@ -0,0 +1,60 @@
+  public WinningInfo getWinningInfo(LottoWinningNumber lottoWinningNumber) {
+    Map<LottoRank, Integer> winningInfo = new HashMap<>();
+
+    for (LottoTicket lottoTicket : lottoTickets) {
+      int matchedCount = lottoWinningNumber.getMatchedCount(lottoTicket);
+        if (matchedCount < MIN_MATCH_COUNT) {
+            continue;
+        }
+      boolean matchedBounus = lottoWinningNumber.isContainsBounusNumber(lottoTicket);
+
+      LottoRank rank = LottoRank.of(getRank(matchedCount, matchedBounus));
```
List의 인덱스 값으로 등수를 관리하던 이전 코드를 보면 엄청나게 개선 하신 걸 알 수 있죠. 👍 💯

```java
    public static final int MIN_MATCH_COUNT = 3;

    public WinningInfo getWinningInfo(LottoWinningNumber lottoWinningNumber) {
        List<Integer> winningInfo = new ArrayList<>(Arrays.asList(0, 0, 0, 0, 0));
        for (LottoTicket lottoTicket : lottoTickets) {
            checkLottoTicket(lottoWinningNumber, winningInfo, lottoTicket);
        }

        return new WinningInfo(winningInfo);
    }

    private void checkLottoTicket(LottoWinningNumber lottoWinningNumber, List<Integer> winningInfo,
        LottoTicket lottoTicket) {
        int winningInfoIndex = 0;
        int matchedCount = lottoWinningNumber.getMatchedCount(lottoTicket);

        if (matchedCount < MIN_MATCH_COUNT)
            return;

        if (isSecond(lottoWinningNumber, lottoTicket, matchedCount) || isFirst(matchedCount)) {
            winningInfoIndex = 1;
        }
        updateWinningInfo(winningInfo, winningInfoIndex, matchedCount);
    }

    private void updateWinningInfo(List<Integer> winningInfo, int winningInfoIndex, int matchedCount) {
        winningInfoIndex += matchedCount - MIN_MATCH_COUNT;
        winningInfo.set(winningInfoIndex, winningInfo.get(winningInfoIndex) + 1);
    }

    private boolean isFirst(int matchedCount) {
        return matchedCount == LottoTicket.LOTTO_NUMBERS_LENGTH;
    }

    private boolean isSecond(LottoWinningNumber lottoWinningNumber, LottoTicket lottoTicket,
        int matchedCount) {
        return matchedCount == LottoTicket.LOTTO_NUMBERS_LENGTH - 1 && lottoWinningNumber.isContainsBounusNumber(lottoTicket);
    }
```

사실 List일 땐 첫번째 요소가 일등일지 꽝일지부터 파악하고 로직 분석을 했네요. 😋

Hue9010 reviewed (approved):
안녕하세요. 터너~!! 피드백 반영 잘해주셨네요. 👍
추가 피드백을 남겼는데 다음 단계를 진행하면서 같이 반영해 주시면 될것 같아요. 😉
궁금한 사항은 언제든 질문주세요.
고생하셨습니다!

sumniy commented:
꿀같은 정보들 알려주셔서 감사합니다!! 👍👍👍 다음 단계에 적용해보겠습니다!

kakao-zero-1 committed:
docs: 로또 게임 기능 요구사항 목록 추가

kakao-zero-1 committed:
test: 주어진 금액에 맞는 로또 티켓 생성 테스트 추가

kakao-zero-1 committed:
feat: 로또 티켓 클래스 구현

kakao-zero-1 committed:
test: 지난 주 당첨 번호를 입력받아 저장하는 TargetLotto 클래스 테스트 추가

kakao-zero-1 committed:
feat: 타겟 로또의 정보를 담는 클래스 구현

kakao-zero-1 committed:
test: 1부터 45까지의 숫자를 랜덤으로 생성하는 테스트 코드 구현

kakao-zero-1 committed:
feat: 1부터 45까지의 숫자를 랜덤으로 생성하는 클래스 구현

kakao-zero-1 committed:
test: 로또 게임을 진행하는 테스트 코드 구현

kakao-zero-1 committed:
feat: 금액에 맞는 로또 티켓을 생성하고 총 수익을 계산하는 클래스 구현

kakao-zero-1 committed:
feat: 로또의 결과를 담는 클래스 구현

kakao-zero-1 committed:
feat: 로또 번호의 생성 전략을 전달하기 위한 인터페이스 추가

kakao-zero-1 committed:
feat: 전체 로또 게임을 실행하기 위한 I/O view 및 컨트롤러 구현

kakao-zero-1 committed:
refactor: 코드 리팩토링

nimunsang committed:
refactor: 코드 스타일 Default로 변경

nimunsang committed:
refactor: 변수명 및 메서드명 리팩토링

nimunsang committed:
refactor: 인터페이스 구현체에 override 어노테이션 추가

sah3122 reviewed (changes requested):
안녕하세요 😄 
로또 미션 리뷰를 맡게된 이동철입니다. 

로또 미션 구현잘해주셨습니다 👍 
몇가지 코멘트를 남겨두었는데 확인 부탁드려요 🙏

nimunsang committed:
refactor: 메서드명을 의미에 맞게 변경

nimunsang committed:
feat: 로또 번호 클래스 선언

- 1~45까지의 번호를 보장
- 객체를 재사용하기 위한 캐시 구현
- 정렬을 위한 comparable 구현
- 생성 및 유효성 검증에 대한 테스트 코드 작성

nimunsang committed:
feat: 로또 번호들을 담는 일급콜렉션 구현

- 중복되는 숫자에 대한 유효성 검사 수행
- 다른 로또 번호들과 몇 개의 숫자가 일치하는지 검사

nimunsang committed:
feat: TargetLotto의 보너스 번호 유효성 검증 및 리팩토링

- 중복되는 번호인지 유효성 검증 수행
- LottoNumbers, LottoNumber로 구성

nimunsang committed:
refactor: LottoTicket 대신 LottoNumbers 사용

nimunsang committed:
test: 테스트 코드 보완

nimunsang committed:
refactor: 메서드의 역할에 맞는 네이밍으로 변경

nimunsang committed:
refactor: 로또번호 생성 시 Integer 대신 LottoNumber 반환

nimunsang committed:
feat: LottoNumber를 추가하며 필요한 메서드 작성

- LottoGame에서 돈의 유효성 검사 추가
- LottoRank, LottoResult에 getter 추가
- LottoTickets에서 TargetLotto와의 비교하는 로직 수행

sah3122 reviewed (approved):
피드백 반영 잘해주셨습니다 👍 
사소한 코멘트 남겨두었는데 다음 미션도 화이팅 해주세요 😄
